ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -194,12 +194,19 @@ func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request)
 			err = r.updateState(ctx, rayJobInstance, nil, rayJobInstance.Status.JobStatus, rayv1alpha1.JobDeploymentStatusWaitForDashboard, err)
 			return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, err
 		}
+		// Check the dashboard readiness by checking the err return from rayDashboardClient.GetJobInfo.
+		// Note that rayDashboardClient.GetJobInfo returns no error in the case of http.StatusNotFound.
```

Comment:
> Note that rayDashboardClient.GetJobInfo returns no error in the case of http.StatusNotFound.

A bit confused by this comment -- this means if we get http.StatusNotFound we'll proceed with the job submission? When does this situation happen?

Reply:
> When does this situation happen?

`http.StatusNotFound` is 404. We should not proceed with the job submission if the dashboard URL is 404 not found.

https://github.com/ray-project/kuberay/blob/92651545d9730904ae9cbe40c5c39e5803e6bb9d/ray-operator/controllers/ray/utils/dashboard_httpclient.go#L354-L356

Reply:
I see, so it sounds like the PR in its current state needs to be updated. 

Should we update the dashboard client to return an error in the case of 404, or should we check if the response is 404 on the job controller side?  

My guess is that updating it in the job controller will have fewer unexpected side effects, but it still seems a bit weird that the HTTP client returns `nil, nil` in this case.

Reply:
> Should we update the dashboard client to return an error in the case of 404?

This seems to be more make sense to me. What do you think?

Reply:
Yeah I think this makes more sense.  We should just check that it doesn't break anything for other callers, but I don't think it will

Reply:
I think it is ok for GetJobInfo to return nil, nil since callers can check the NotFound situation by checking the first return value is nil or not, and they should always do.

With that in mind, returning NotFound as an error seems a little bit redundant.

Reply:
> I think it is ok for GetJobInfo to return nil, nil since callers can check the NotFound situation by checking the first return value is nil or not, and they should always do.

Yes, users can examine the first returned value, but it seems somewhat irregular not to throw an error when the retrieval of job information fails.

Reply:
Sure. Just note that some people may consider the NotFound as a successful result and I believe that is the reason why the original author chose to return nil err in the GetJobInfo.

Should I change its behavior in this PR? Or should I open another issue for it first?

Reply:
> Just note that some people may consider the NotFound as a successful result and I believe that is the reason why the original author chose to return nil err in the GetJobInfo.

If the function is `DeleteJobInfo`, it makes sense to interpret a 404 as a successful result, as the job may have been deleted by the previous reconciliation. However, I cannot conceive of any justification for deeming a 404 in GetJobInfo as success. Can you think of any specific reasons?

> Should I change its behavior in this PR? Or should I open another issue for it first?

I prefer not to include the change in this PR. We can open an issue to track the progress and discussion.

Reply:
> However, I cannot conceive of any justification for deeming a 404 in GetJobInfo as success. Can you think of any specific reasons?

It is just another standpoint of designing API behavior, regardless of API naming, saying "Hey, I have finished what you asked for successfully without any error. And the result is no result".